[libc++] Fix copy-paste damage in `ranges::rotate_copy` and its test

StephanTLavavej opened the pull request:
Found while running libc++'s tests with MSVC's STL.

`ranges::rotate_copy` takes `forward_iterator`s as this test's comment banner correctly depicts. However, this test had bogus assertions expecting that `ranges::rotate_copy` would be constrained away for not-quite-**bidi** iterators. @philnik777 confirmed that these were copy-paste relics from the `ranges::reverse_copy` test.

I fixed this by replacing the assertions with the test types that aren't quite **forward** iterators/ranges. Additionally, I noticed that the top-level `test()` function was missing coverage with the weakest possible `forward_iterator<int*>`.

This revealed that the product code in `ranges_rotate_copy.h` was similarly damaged. In addition to fixing it by taking `forward_iterator` and `forward_range` as depicted in the Standard, this drops the inclusion of `<__iterator/reverse_iterator.h>` as this algorithm doesn't need `std::__reverse_range`.

StephanTLavavej committed:
Fix copy-paste damage in ranges.rotate_copy.pass.cpp.

CaseyCarter commented:
> Note: The fact that this test was passing for you is quite curious. From a quick glance, I don't see a problem with the libc++ product code constraints, so I'm not sure where the behavior difference is.

Try restoring the original `static_assert`s but negating their expected results. The failure diagnostics will probably be illuminating.

StephanTLavavej commented:
Good idea - which I didn't even need to try, as extending the test coverage with `forward_iterator<int*>` revealed that the product code was also copy-pasted saying `bidirectional`:

https://github.com/llvm/llvm-project/blob/06c5c27e44757e4917f7d7c995720ae8a2d3b1f9/libcxx/include/__algorithm/ranges_rotate_copy.h#L35-L56

Let's see if I can fix this despite not being tooled up to work on libc++ product code.

StephanTLavavej committed:
Fix copy-paste damage in ranges_rotate_copy.h.

In addition to taking `forward_iterator` and `forward_range`
as depicted in the Standard, this drops the inclusion of
`<__iterator/reverse_iterator.h>` as this algorithm
doesn't need `std::__reverse_range`.

philnik777 reviewed (approved):
Thanks for the fix! LGTM with green CI.

StephanTLavavej commented:
Thanks, CI is green with only one spurious build failure on Apple. Merging now.
